test: Add end-to-end tests for UI behavior

rahulsom opened the pull request:
I want to remove the dependency on Prototype so this plugin is compatible with Jenkins 2.401 and above. Before I do that, I want to write an end-to-end test.

See [this post](https://www.jenkins.io/blog/2023/05/12/removing-prototype-from-jenkins/) for more details on the removal of Prototype.

https://issues.jenkins.io/browse/JENKINS-71365 is the issue for this plugin.
https://github.com/jenkinsci/stapler/pull/452/files shows changes that have been made to stapler which is a dependency of core.

<!-- Please describe your pull request here. -->

### Testing done

This is a pure test PR. It adds a dependency on `org.seleniumhq.selenium:selenium-java` and relies on `ChromeDriver` to run the actual test. The test works in both headless and normal mode.

The test checks for all 4 kinds of params - `SINGLE_SELECT`, `RADIO`, `CHECKBOX`, `MULTI_SELECT`.
It creates `ChoiceParameter` and a `CascadeChoiceParameter` for each kind to check the UI works.

```[tasklist]
### Submitter checklist
- [x] Make sure you are opening from a **topic/feature/bugfix branch** (right side) and not your main branch!
- [x] Ensure that the pull request title represents the desired changelog entry
- [x] Please describe what you did
- [x] Link to relevant issues in GitHub or Jira
- [x] Link to relevant pull requests, esp. upstream and downstream changes
- [x] Ensure you have provided tests - that demonstrates feature works or fixes the issue
```

<!--
Put an `x` into the [ ] to show you have filled the information.
The template comes from https://github.com/jenkinsci/.github/blob/master/.github/pull_request_template.md 
You can override it by creating .github/pull_request_template.md in your own repository 
-->

rahulsom commented:
I initially picked the latest version `4.10.0` of `selenium-java` which supports Chrome 114.
The Jenkins test infrastructure supports only Chrome 112.
The relationship between these can be fragile.

I've tried pinning back the library a few versions `4.8.0` - to no avail.
I tried using [WebDriverManager](https://github.com/bonigarcia/webdrivermanager) - that didn't help much either.

I'll use these tests to make my actual change and test it manually. If you feel like you want to give this a shot, feel free to resurrect it. I'm inclined to close this PR.

kinow commented:
Maybe cypress or nightwatch would work too? (I use Cypress in a few projects). Thanks for the PR!

rahulsom commented on src/test/java/org/biouno/unochoice/UiAcceptanceTest.java:
```diff
@@ -0,0 +1,200 @@
+
+    @BeforeClass
+    public static void setUpClass() {
+        WebDriverManager.chromedriver().browserVersion("112").setup();
```
I had to put `.browserVersion("112")` here for this to work on Jenkins infrastructure.
On my laptop, I can remove it and it picks up `114` automatically.

This is not ideal, but it is a helpful validation. If it feels hacky, we can close this PR.

rahulsom commented:
> Maybe cypress or nightwatch would work too? (I use Cypress in a few projects). Thanks for the PR!

I think this test needs the front-end and back-end both spun up. The tooling that exists in Jenkins for running tests in Java seems better suited at the moment.

There is some [ongoing work](https://github.com/sghill/testcontainers-jenkins) for spinning a Jenkins instance with the plugin in a container - that could be helpful for running the validation in Cypress/Nightwatch.

kinow reviewed (commented):
>I think this test needs the front-end and back-end both spun up. The tooling that exists in Jenkins for running tests in Java seems better suited at the moment.

Ah, right, that does sound like using Jenkins tooling will be easier/better :+1: 

>There is some [ongoing work](https://github.com/sghill/testcontainers-jenkins) for spinning a Jenkins instance with the plugin in a container - that could be helpful for running the validation in Cypress/Nightwatch.

:+1: good to know!

@rahulsom just to confirm because I think I saw you were working on this branch a few days ago. Is this one ready to be reviewed & merged?

Thank you!

rahulsom commented on src/test/java/org/biouno/unochoice/UiAcceptanceTest.java:
```diff
@@ -0,0 +1,214 @@
+    public static void setUpClass() {
+        if (isCi()) {
+            // The browserVersion needs to match what is provided by the Jenkins Infrastructure
+            WebDriverManager.chromedriver().browserVersion("112").setup();
```
Yep. This is the place we would need to update the version.
The documentation on https://bonigarcia.dev/webdrivermanager/#motivation seems to suggest this should be automatic.

I am not sure how to find it from Jenkins Infra - It's an OS dependency.
We can wait to see how bothersome this is, and disable it on Jenkins Infra if this changes often.

[JUnit's Assume](https://junit.org/junit4/javadoc/4.12/org/junit/Assume.html) should help with that. We could do a 
```java
Assume.assumeFalse(isCi());
```

rahulsom replied:
When it fails for a version mismatch, this is the exception

```
	org.openqa.selenium.SessionNotCreatedException: Could not start a new session. Response code 500. Message: session not created: This version of ChromeDriver only supports Chrome version 114
Current browser version is 112.0.5615.49 with binary path /usr/bin/chromium-browser 
```

kinow replied:
All good, can you add a comment like "if you see the error SessionNot... etc etc, it means you must update this version....". Just so if that happens I can try to troubleshoot it myself. After that, LGTM :rocket: ! Thanks!

basil replied:
I'm doing a similar working in https://github.com/jenkinsci/build-monitor-plugin/blob/7a2a3d24656d11aca23be8d7cd716970b99023c0/build-monitor-acceptance/pom.xml#L141 by setting the `wdm.chromeDriverVersion`. As with this plugin, the workaround does not work on Windows.

I can't help but think this should just work out of the box on our container agents — at least the Linux ones (the Windows ones, as I understand it, are currently built using a legacy system that is prohibitively difficult to work on and that the infrastructure team intends to get rid of). If we can distill this use case into a concrete ask for the infrastructure team, I am sure they could update the Linux images as requested.

rahulsom commented:
Ran into some docker weirdness for a bit - chrome was using an in-memory store and crashing. 
I've updated the test to be more resilient with that - it now uses the `/tmp` volume.
I've also rebased against master so the merge is less noisy in the commit history.

kinow reviewed (approved):
Thanks!
